airbyte-integrations/bases/base-typing-deduping/src/main/java/io/airbyte/integrations/base/destination/typing_deduping/TypeAndDedupeOperationValve.java

```diff
@@ -0,0 +1,104 @@
+package io.airbyte.integrations.base.destination.typing_deduping;
+
+import io.airbyte.protocol.models.v0.AirbyteStreamNameNamespacePair;
+
+import java.util.List;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.function.Supplier;
+
+/**
+ * A slightly more complicated way to keep track of when to perform type and dedupe operations per stream
+ */
+public class TypeAndDedupeOperationValve extends ConcurrentHashMap<AirbyteStreamNameNamespacePair, Long> {
+
+    private static final long TWO_MINUTES_MILLIS = 1000 * 60 * 2;
+
+    private static final long FIVE_MINUTES_MILLIS = 1000 * 60 * 5;
+
+    private static final long TEN_MINUTES_MILLIS = 1000 * 60 * 10;
+
+    private static final long FIFTEEN_MINUTES_MILLIS = 1000 * 60 * 15;
+
+    // New users of airbyte likely want to see data flowing into their tables as soon as possible
+    // However, as their destination tables grow in size, typing and de-duping data becomes an expensive operation
+    // To strike a balance between showing data quickly and not slowing down the entire sync, we use an increasing
+    // interval based approach. This is not fancy, just hard coded intervals.
+    private static final List<Long> typeAndDedupeIncreasingIntervals = List.of(
```

Comment:
These are somewhat arbitrary, open to changing them or making them configurable based on sync type?